# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [ ]:
#Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"
#4 would be the argument 
#wrap lambda to execute

(lambda num : num + 2)(4)

In [ ]:
#can take multiple inputs

#see below

In [23]:
def addTwo(num): #num is a variable name
    return num + 2

print(addTwo(5))

7


#### Saving to a Variable

In [1]:
addTwoLambda = lambda num: num + 2

addTwoLambda(10)

12

#### Multiple Inputs

In [24]:
(lambda num1, num2, num3: num1 + num2)(6,4,10)

10

In [3]:
#store as a variable

addThreeNumbers = lambda num1, num2, num3: num1 + num2 + num3

addThreeNumbers(6,4,10)

20

In [4]:
#regular function to compare

def addThreeFunc(num1, num2, num3):
    return num1 + num2 + num3

#### Passing a Lambda into a Function

In [17]:
#first find function

def math_operations(f, num):
    return f(num) #define f by passing in our argument
math_operations(lambda x: x**2, 10) #takes in single input num or x, take lambda and square it, 2nd arg is 10 =100

math_operations(lambda num: num//10, 1000) #first function is a lambda one
#-------
multiply_by_ten = lambda num: num * 10 #stored not executed yet

math_operations(multiply_by_ten, 5) #=50

50

35

#### Returning a Lambda from a Function

In [26]:
#function with a nested function, will return nested function
#write out in 2 functions first

def outterFunction():
    outter_num=4
#print(inner_num)Error cant access variables from nested function    
    def multiply(num):
        inner_num = 10
        return num * outter_num
    return multiply #then run the return #2


#execute outter function
print(outterFunction()(5)) #run the print #1

my_multiply = outterFunction()

my_multiply(4)

20


16

In [ ]:
def returnLambda(num1,num2):
    return lambda num3, num4 : num1 + num2 + num3 + num4 #access outter paramaeter then innter parameters

addTwoThree = returnLambda(2,3)
#print(addTwoThree)
addTwoThree(10,20)

In [28]:
#another ex

def how_to_multiply(num1):
    return lambda num2 : num1 * num2

double = how_to_multiply(2)

triple = how_to_multiply(3)

print(double(10))

triple(10)

20


30

#### If Statements within Lambdas

In [31]:
#if requires an else

greater_than_10_add_2 = lambda num : num + 2 if num > 10 else num #include else

greater_than_10_add_2(3)

3

#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [21]:
#def math_operations(f, num):
    #return f(num) #define f by passing in our argument
#math_operations(lambda x: x**2, 10) #takes in single input num or x, take lambda and square it, 2nd arg is 10 =100

def math_operations (f , num):
    return f(num)
math_operations(lambda num: num **3 , 2)

f = ;a,nda num: num**2 if num%2
    
    #lambda <parameter>: <Code to return>

100

## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [24]:
#map(function to be used, list to be used)
#map(<function>, iterable)
#will return map object

def square_number(num):
    return num**2

number_list = [1,2,3,4,5,6]

#to perform square wlwmwnt in each element in the brackets
#list function
list(map(square_number, number_list)) #lambaa, map, then list

[1, 4, 9, 16, 25, 36]

#### Using Lambda's with Map

In [33]:
number_list = [1,2,3]
for num in map(lambda num: num**3, number_list): #will return on whats on right side of colon, first argument (), then pass in illst
    print(num, end=' ')
    
#same code with list comp    
[num for num in map(lambda num: num**3, number_list)]

1 8 27 

[1, 8, 27]

In [34]:
attackers = [5,6,2,8]
defenders = [4,7,1,7]

survivors = list(map(lambda attacker, defender: attacker < defender, attackers, defenders)) #output: [True, False, True, True]
print(survivors.count(True), ' ', survivors.count(False))
survivors.count(True) > survivors.count(False)

1   3


False

In [ ]:
#map(<function>, iterable) returns a map object

#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [46]:
double_list = [4, 7, 12, 19]
double_minus_one = list(map(lambda num: num*2 - 1, double_list))
print(double_minus_one)

[7, 13, 23, 37]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [50]:
# filter(<function>, iterable) returns filter

def greater_than_nine(num):
    

#same as: 
#    if num >=10:
#        return true
#    else: 
#        return false

    return num >= 10

num_list = [10,1,100,5,11]

list(filter(greater_than_nine, num_list)) #if false, will filter out element, we dont want it

[10, 100, 11]

#### Using Lambda's with Filter()

In [3]:
students = ['Hamed', 'Andy', 'William', 'Chris', 'Will']

#filter(lambda student: student[0].lower == 'w', students) #comparing 1st letter of stud, indexing into string, students is iterable

[student for student in filter(lambda student: student[0].lower() == 'w', students)]


['William', 'Will']

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [8]:
import statistics

random_numbers = [2,7,4.2,1.6,9,4.4,4.9]

avg = statistics.mean(random_numbers)

[num for num in filter(lambda num: num < avg, random_numbers)]

#output = [2, 4.2, 1.6, 4.4]

[2, 4.2, 1.6, 4.4]

## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [34]:
#must import

from functools import reduce


#reduce(<function>, iterable)

print(sum([1,2,3,4,5])) #reduced to one number output being 6 so 1+2. = 3 +3 = 6

def add_two_numbers(num1,num2):
    return num1+num2

reduce(add_two_numbers, [1,2,3,4,5])

15


15

#### Using Lambda's with Reduce()

In [11]:
#instead of adding we will multiply

reduce(lambda num1,num2 : num1*num2, [1,2,3,4,5]) #names our parameters then go thru list

120

#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [12]:
#subtract every number

from functools import reduce

reduce(lambda num1,num2 : num1-num2, [1,2,3,4,5])


-13

## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [18]:
def add_num(num):
    if num == 1:
        print(f'add_num({num})=1')
        return 1
    print(f'add_num({num})={num} * add_num{num - 1}')
    return num + add_num(num - 1)

add_num(5)

add_num(5)=5 * add_num4
add_num(4)=4 * add_num3
add_num(3)=3 * add_num2
add_num(2)=2 * add_num1
add_num(1)=1


15

#### Writing a Factorial Function

In [16]:
def find_factorial(num):
    if num == 1:
        print(f'find_factorial({num})=1')
        return num
    print(f'find_factorial({num})={num} * find_factorial{num - 1}')
    return num * find_factorial(num -1)

find_factorial(5)

find_factorial(5)=5 * find_factorial4
find_factorial(4)=4 * find_factorial3
find_factorial(3)=3 * find_factorial2
find_factorial(2)=2 * find_factorial1
find_factorial(1)=1


120

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [40]:
def find_factorial(num):
    if num == 1:
        print(f'find_factorial({num})=1')
        return num
    print(f'find_factorial({num})={num} - find_factorial{num - 1}')
    return num - find_factorial(num-1)
find_factorial(4)

find_factorial(4)=4 - find_factorial3
find_factorial(3)=3 - find_factorial2
find_factorial(2)=2 - find_factorial1
find_factorial(1)=1


2

In [22]:
def big_nums(num):
    if num == 1:
        return 1
    return num ** big_nums(num - 1)

big_nums(4)

262144

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [53]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

filter_strings = list(filter(lambda s: s.strip() != "", places))

print(filter_strings)

['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [72]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J.Bernstein"]
#lambda sorts last name, use sorted with optional 'key' argument

sorted_last_name = sorted(author, key=lambda n: n.split()[-1].lower())

print(sorted_last_name)

['Gary A.J.Bernstein', 'Victor aNisimov', 'Joel Carter', 'David hassELHOFF', 'Andrew P. Garfield']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [60]:
# F = (9/5)*C + 32

places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

#loop thru each element and recursion for each tuple use F

c_to_f = list(map(lambda t: (t[0], (9/5)*t[1]+32), places))

print(c_to_f)

[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [66]:
#looking for a lambda, use recursion
#will get video on sequence

def fibonacci(n):
    a, b = 1, 1
    for i in range(n):
        a, b = b, a+b
    return a
for i in range(6):
    print(fibonacci(i))

1
1
2
3
5
8
